driver/ixsystems/common.py

```diff
@@ -697,3 +696,67 @@ def replication_run(self, id):
             LOG.debug(f'replication_run response: {rep}')
         except FreeNASApiError as api_error:
             raise FreeNASApiError('Unexpected error', api_error) from api_error
+
+    def get_volume(self, volume_id):
+        """ Use dataset API /v2.0/pool/dataset/id/{id} to get volume details
+        """
+        LOG.debug(f'get_volume {volume_id}')
+        encoded_datapath = urllib.parse.quote_plus(
+                self.configuration.ixsystems_dataset_path + '/' + volume_id)        
+        request_urn = (f"/pool/dataset/id/{encoded_datapath}/")
+        try:
+            rep = self.handle.invoke_command(
+                FreeNASServer.SELECT_COMMAND,
+                request_urn, None)
+            LOG.debug(f'get_volume response: {rep}')
+            represult = json.loads(rep['response'])  
+            return represult        
+        except FreeNASApiError as api_error:
+            raise FreeNASApiError('Unexpected error', api_error) from api_error
+
+    def get_snapshot(self, volume_name, snapshot_name):
+        """ Use dataset API /v2.0/zfs/snapshot/id/{id} to get snapshot details
+        """
+        LOG.debug(f'get_snapshot {volume_name}@{snapshot_name}')
+        encoded_datapath = urllib.parse.quote_plus(
+                self.configuration.ixsystems_dataset_path + '/' 
+                + volume_name) + '@' + snapshot_name
+        request_urn = (f"/zfs/snapshot/id/{encoded_datapath}/")
+        try:
+            rep = self.handle.invoke_command(
+                FreeNASServer.SELECT_COMMAND,
+                request_urn, None)
+            LOG.debug(f'get_snapshot response: {rep}')
+            represult = json.loads(rep['response'])  
+            return represult        
+        except FreeNASApiError as api_error:
+            raise FreeNASApiError('Unexpected error', api_error) from api_error
+
+    def get_all_snapshot(self):
+        """ Use zfs snapshot API /v2.0/zfs/snapshot/ to get all snapshot details
+        """
+        LOG.debug(f'get_all_snapshot')
+        request_urn = (f"/zfs/snapshot/")
+        try:
+            rep = self.handle.invoke_command(
+                FreeNASServer.SELECT_COMMAND,
+                request_urn, None)
+            LOG.debug(f'get_all_snapshot response: {rep}')
+            represult = json.loads(rep['response'])  
+            return represult        
+        except FreeNASApiError as api_error:
+            raise FreeNASApiError('Unexpected error', api_error) from api_error
+
+    def get_volume_from_snapshot(self, snapshot_name):
+        """
+        Use zfs snapshot API /v2.0/zfs/snapshot/ to get volume name from snapshot
+        """
+        all_snapshot = self.get_all_snapshot()
```

Comment:
To avoid get_all_snapshot, is there something else we could do here?
What is the format of snapshot_name? tank/foo@name? or just name?

We could try to get the specific snapshot from all datasets, depending on the case.